### PR TITLE
=str 20711 IllegalMonitorStateException in Source.queue

### PR DESCRIPTION
**Pull request by agolubev**
Fix is that after stop: events will handle sequentially

### CODE REVIEW

**Comment by ktoso** [comment]
OK TO TEST

**Comment by ktoso** [comment]
Refs #20711

**Comment by agolubev** [comment]
Build failure not related:

(akka-bench-jmh/compile:compileIncremental) Compilation failed

**Comment by ktoso** [comment]
PLS BUILD

**Comment by ktoso** [comment]
LGTM AFAICS

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 1]
```diff
@@ -1309,14 +1309,12 @@ private[akka] trait CallbackWrapper[T] extends AsyncCallback[T] {
...
-      case Stopped(cb) ⇒
-        lock.unlock()
-        cb(arg)
+      case Stopped(cb)              ⇒ cb(arg)
     }
   }
 
   private[this] def locked(body: ⇒ Unit): Unit = {
     lock.lock()
-    try body finally if (lock.isLocked) lock.unlock()
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: This method isn't used anywhere?

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Damn, I'm blind, now I see. NVM!

Comment (ktoso) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Yup, it wraps all callbacks

File: akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala [hunk 2]
```diff
@@ -1309,14 +1309,12 @@ private[akka] trait CallbackWrapper[T] extends AsyncCallback[T] {
...
+    try body finally lock.unlock()
```
Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: So, how could this cause the exception in the ticket, didn't that say that there was an `unlock` call to an unlocked lock? So the invoke lock block completed with the lock unlocked inside of it, then the if clause would not match and the lock wouldn't be double-unlocked and cause the exception? What am I missing here?

Comment (agolubev) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: @johanandren  you are correct. There can be can be another unlock  between isLocked and unlock that makes fail the second one.

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Thanks for clarifying

Comment (johanandren) [inline akka-stream/src/main/scala/akka/stream/stage/GraphStage.scala]: Ah, it was described in the ticket even. Sorry about that.

**Comment by johanandren** [comment]
LGTM!

**Pull request merged**